### PR TITLE
[LLVM] Add InsertPosition union-type to remove overloads of Instruction-creation

### PR DESCRIPTION
**Pull request by SLTozer**
This patch follows the previous PR https://github.com/llvm/llvm-project/pull/94224 and contains that patch's commit; to view the changes for only this patch, review only the [second commit](https://github.com/llvm/llvm-project/pull/94226/commits/ffd3172798a84396c8b535d527c8c4ab1597d87b).

This patch builds off the previous PR to simplify instruction creation, by replacing all overloads of instruction constructors/Create methods that are identical other than the `Instruction *InsertBefore`/`BasicBlock *InsertAtEnd`/`BasicBlock::iterator InsertBefore` argument with a single version that takes an `InsertPosition` argument. The `InsertPosition` class can be implicitly constructed from any of the above, internally converting them to the appropriate `BasicBlock::iterator` value which can then be used to insert the instruction (or to not insert it if an invalid iterator is passed).

The upshot of this is that code will be deduplicated, and all callsites will switch to calling the new unified version without any changes needed to make the compiler happy. There is at least one exception to this; the construction of `InsertPosition` is a user-defined conversion, so any caller that was already relying on a different user-defined conversion won't work. In all of LLVM and Clang this happens exactly once: at `clang/lib/CodeGen/CGExpr.cpp:123` we try to construct an alloca with an `AssertingVH<Instruction>` argument, which must now be explicitly cast to an `Instruction*`. If this is more common elsewhere, it could be fixed by adding an appropriate constructor to `InsertPosition`.

The end result of this is [performance-neutral](http://llvm-compile-time-tracker.com/compare.php?from=b2bd024384b484647da9fd9863bf6f77b5731949&to=3950bc6991675714782e05f732ae4b1ce7b6bc25&stat=instructions%3Au), with no significant impact on the geomean performance (albeit with some specific projects seeing a compile time increase), and a slight reduction in clang build time and file size. Besides reducing line count heavily and simplifying code, this helps build towards [the proposal to rewrite the instruction-insertion interface](https://discourse.llvm.org/t/rfc-rewriting-the-instruction-insertion-interface/78494/1), which will help reduce future errors relating to incorrect instruction insertion in the presence of debug records. Getting this single large change in now will also make it much easier to deprecate the `Instruction*` insertion arguments, [as proposed on discourse](https://discourse.llvm.org/t/psa-instruction-constructors-changing-to-iterator-only-insertion/77845/1), as we can do so with a single `LLVM_DEPRECATED` tag rather than needing to update every single method.

### CODE REVIEW

**Comment by SLTozer** [comment]
Abandoned in favour of https://github.com/llvm/llvm-project/pull/94226.

File: clang/lib/CodeGen/CGExpr.cpp [hunk 1]
```diff
@@ -120,7 +120,8 @@ llvm::AllocaInst *CodeGenFunction::CreateTempAlloca(llvm::Type *Ty,
     Alloca = Builder.CreateAlloca(Ty, ArraySize, Name);
   else
     Alloca = new llvm::AllocaInst(Ty, CGM.getDataLayout().getAllocaAddrSpace(),
-                                  ArraySize, Name, AllocaInsertPt);
+                                  ArraySize, Name,
+                                  (llvm::Instruction *)AllocaInsertPt);
```
Comment (nikic) [inline clang/lib/CodeGen/CGExpr.cpp:124]:
```suggestion
                                  &*AllocaInsertPt);
```
Does something like this work instead?

File: llvm/include/llvm/IR/IRBuilder.h [hunk 2]
```diff
@@ -63,10 +63,10 @@ class IRBuilderDefaultInserter {
   virtual ~IRBuilderDefaultInserter();
 
   virtual void InsertHelper(Instruction *I, const Twine &Name,
-                            BasicBlock *BB,
                             BasicBlock::iterator InsertPt) const {
-    if (BB)
-      I->insertInto(BB, InsertPt);
+    if (InsertPt.isValid()) {
```
Comment (nikic) [inline llvm/include/llvm/IR/IRBuilder.h:67]: Drop braces.

File: llvm/include/llvm/IR/IRBuilder.h [hunk 3]
```diff
@@ -63,10 +63,10 @@ class IRBuilderDefaultInserter {
...
+      I->insertInto(InsertPt.getNodeParent(), InsertPt);
```
Comment (nikic) [inline llvm/include/llvm/IR/IRBuilder.h:68]: We should drop the BasicBlock argument from this API now, but that can be a followup...

File: llvm/include/llvm/IR/Instruction.h [hunk 4]
```diff
@@ -44,6 +44,23 @@ template <> struct ilist_alloc_traits<Instruction> {
...
+public:
+  InsertPosition(std::nullopt_t) : InsertAt() {}
+  InsertPosition(std::nullptr_t) : InsertAt() {}
+  // LLVM_DEPRECATED("Use BasicBlock::iterators for insertion instead",
+  // "BasicBlock::iterator")
+  InsertPosition(Instruction *InsertBefore);
+  InsertPosition(BasicBlock *InsertAtEnd);
+  InsertPosition(InstListType::iterator InsertAt) : InsertAt(InsertAt) {}
+  operator InstListType::iterator() const { return InsertAt; }
+  bool IsValid() const { return InsertAt.getNodePtr(); }
```
Comment (nikic) [inline llvm/include/llvm/IR/Instruction.h:61]: I think you added an isValid() method to the iterator itself?

File: llvm/lib/IR/Instructions.cpp [hunk 5]
```diff
@@ -1460,56 +1184,29 @@ static Value *getAISize(LLVMContext &Context, Value *Amt) {
   return Amt;
 }
 
-static Align computeAllocaDefaultAlign(Type *Ty, BasicBlock *BB) {
-  assert(BB && "Insertion BB cannot be null when alignment not provided!");
+static Align computeAllocaDefaultAlign(Type *Ty, InsertPosition Pos) {
+  assert(Pos.IsValid() &&
+         "Insertion position cannot be null when alignment not provided!");
+  BasicBlock *BB = ((BasicBlock::iterator)Pos).getNodeParent();
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:1190]: Maybe add a getParent() method on InsertPosition?

**Review by nikic — APPROVED** [review]
Thanks for doing this, LGTM!

**Comment by nikic** [comment]
It looks like polly needs an update as well.

File: llvm/lib/IR/Instructions.cpp [hunk 6]
```diff
@@ -1567,68 +1242,33 @@ void LoadInst::AssertOK() {
          "Ptr must have pointer type.");
 }
 
-static Align computeLoadStoreDefaultAlign(Type *Ty, BasicBlock *BB) {
-  assert(BB && "Insertion BB cannot be null when alignment not provided!");
+static Align computeLoadStoreDefaultAlign(Type *Ty, InsertPosition Pos) {
+  assert(Pos.isValid() &&
+         "Insertion position cannot be null when alignment not provided!");
+  BasicBlock *BB = ((BasicBlock::iterator)Pos).getNodeParent();
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:1248]:
```suggestion
  BasicBlock *BB = Pos.getBasicBlock();
```

**Comment by SLTozer** [comment]
Alright, builds now working for LLVM, Clang, Flang, Polly, and MLIR; looks good to merge?

**Review by nikic — APPROVED** [review]

**Comment by efriedma-quic** [comment]
I think some of the overloads for constructing an instruction aren't quite right: in some cases, we require a non-null insertion-point so we can retrieve the DataLayout from the insertion-point.  (Particularly instructions that have an explicitly specified alignment.)  In those cases, we shouldn't default to "= nullptr": a null insertion point is guaranteed to assert at runtime.

**Comment by SLTozer** [comment]
> I think some of the overloads for constructing an instruction aren't quite right:

That's reasonable - I defaulted to adding nullptr-defaults in the case where it didn't cause incorrect overloads, but it's easier to keep only the existing defaults and let any others be added later as appropriate; I've updated the patch accordingly.

**Review by efriedma-quic — APPROVED** [review]

**Pull request merged**